src/algebra/monoid_algebra/degree.lean

```diff
@@ -0,0 +1,160 @@
+  refine Dm.trans (add_le_add _ _);
+  exact finset.le_sup ‹_›,
+end
+
+end add_only
+
+lemma sup_support_list_prod_le [add_monoid A] [add_monoid B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b) :
+  ∀ F : list (add_monoid_algebra R A),
+    F.prod.support.sup D ≤ (F.map (λ f : add_monoid_algebra R A, f.support.sup D)).sum
+| [] := begin
+    rw [list.map_nil, finset.sup_le_iff, list.prod_nil, list.sum_nil],
+    exact λ a ha, by rwa [finset.mem_singleton.mp (finsupp.support_single_subset ha)]
+  end
+| (f::fs) := begin
+    rw [list.prod_cons, list.map_cons, list.sum_cons],
+    exact (sup_support_mul_le Dm _ _).trans (add_le_add_left (sup_support_list_prod_le _) _)
+  end
+
+end semiring
+
+lemma sup_support_pow_le [comm_semiring R] [add_monoid A] [add_monoid B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b) (n : ℕ)
+  (f : add_monoid_algebra R A) :
+  (f ^ n).support.sup D ≤ n • (f.support.sup D) :=
+begin
+  induction n with n hn,
+  { simp only [pow_zero, zero_smul, finset.sup_le_iff],
+    exact λ a ha, by rwa finset.mem_singleton.mp (finsupp.support_single_subset ha) },
+  { rw [pow_succ, succ_nsmul],
+    exact (sup_support_mul_le Dm _ _).trans (add_le_add rfl.le hn) }
+end
+
+variables [comm_semiring R] [add_comm_monoid A] [add_comm_monoid B]
+  [covariant_class B B (+) (≤)] [covariant_class B B (function.swap (+)) (≤)]
+
+lemma sup_support_multiset_prod_le
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b)
+  (F : multiset (add_monoid_algebra R A)) :
+  F.prod.support.sup D ≤ (F.map (λ f : add_monoid_algebra R A, f.support.sup D)).sum :=
+begin
+  induction F using quot.induction_on,
+  rw [multiset.quot_mk_to_coe'', multiset.coe_map, multiset.coe_sum, multiset.coe_prod],
+  exact sup_support_list_prod_le D0 Dm F,
+end
+
+lemma sup_support_finset_prod_le
+  {D : A → B} (D0 : D 0 ≤ 0) (Dm : ∀ a b, D (a + b) ≤ D a + D b)
+  (F : finset (add_monoid_algebra R A)) :
+  (finset.prod F id).support.sup D ≤ finset.sum F (λ f, f.support.sup D) :=
+begin
+  rcases F with ⟨F, hF⟩,
+  rw [finset.prod_mk, multiset.map_id],
+  exact sup_support_multiset_prod_le D0 Dm F,
+end
+
+end general_results_assuming_semilattice_sup
+
+section degrees
+
+variables [semiring R]
+
+section degree
+
+variables [semilattice_sup A]
+
+/--  Let `R` be a semiring, let `A` be a `semilattice_sup`, and let `f : R[A]` be an
+an element of `add_monoid_algebra R A`.  The degree of `f` takes values in `with_bot A` and
+it is the supremum of the support of `f` or `⊥`, depending on whether `f` is non-zero or not.
+
+If `A` has a linear order, then this notion coincides with the usual one, using the maximum of
+the exponents. -/
+def degree (f : add_monoid_algebra R A) : with_bot A :=
```

Comment:
@kmill, this is the type of thing you'd advocate marking
```suggestion
@[reducible]
def degree (f : add_monoid_algebra R A) : with_bot A :=
```
right?

Reply:
I pushed it, but I do not have an opinion on the matter.

Reply:
Also `trailing_degree` would get a `reducible` tag, then, right?

Reply:
Ok, I'll go ahead and mark as `reducible` also `trailing_degree --> min_degree`.